File: src/task12_3_1/Wheel.java
package task12_3_1;

public class Wheel {
    public enum Material {
        Steel, //  стальний диск
        Alloy  //  легкосплавний
    }

    final private Material material;
    final int diameter;

    public Wheel(int diameter) {
        this.material = Material.Steel;
        this.diameter = diameter;
    }

    @Override
    public String toString() {
        return "Wheel{" + this.hashCode() +
                " material=" + material +
                ", diameter=" + diameter +
                '}';
    }
}

File: src/task6_3_2/Wheel.kt
package task6_3_2


class Wheel: Cloneable {

    enum class Shape{
        CIRCLE,
        SQUARE,
        TRIANGLE
    }

    enum class Tire{
        WINTER,
        HIGHWAY,
        SUMMER,
        MUD
    }

    class Builder{

        private var shape: Shape = Shape.CIRCLE
        private var tire: Tire = Tire.HIGHWAY

        fun setShape(shape: Shape): Builder{
            this.shape = shape
            return this
        }

        fun setTire(tire: Tire): Builder{
            this.tire = tire
            return this
        }

        fun build(): Wheel{
            return Wheel(shape, tire)
        }

    }


    private val shape: Shape
    private val tire: Tire

    private constructor(shape: Shape, tire: Tire){
        this.shape = shape
        this.tire = tire
    }

    override fun toString(): String {
        return "${this.hashCode()}-Wheel - shape: $shape; tire: $tire\n"
    }

    override fun copy(): Wheel{
        return Wheel(this.shape, this.tire)
    }

}